zephyr/drivers/can/can_sccan.c

```diff
@@ -0,0 +1,1032 @@
+	struct sc_can_data *data = dev->data;
+	uint32_t idr;
+
+	__ASSERT_NO_MSG(callback != NULL);
+
+	LOG_DBG("Sending %d bytes on %s. Id: 0x%x, ID type: %s %s",
+		frame->dlc, dev->name, frame->id,
+		(frame->flags & CAN_FRAME_IDE) != 0 ? "extended" : "standard",
+		(frame->flags & CAN_FRAME_RTR) != 0 ? ", RTR frame" : "");
+
+	if (frame->dlc > CAN_MAX_DLC) {
+		LOG_ERR("DLC of %d exceeds maximum (%d)",
+			frame->dlc, CAN_MAX_DLC);
+		return -EINVAL;
+	}
+
+	if ((frame->flags & ~(CAN_FRAME_IDE | CAN_FRAME_RTR)) != 0) {
+		LOG_ERR("unsupported CAN frame flags 0x%02x", frame->flags);
+		return -ENOTSUP;
+	}
+
+	/* Check if the TX buffer is full */
+	if (k_sem_take(&data->tx_sem, timeout) != 0) {
+		return -EAGAIN;
+	}
+
+	if (!sc_can_is_enabled(config)) {
+		return -ENETDOWN;
+	}
+
+	k_mutex_lock(&data->inst_mutex, K_FOREVER);
+
+	/* Write CAN IDR to TMR1 */
+	idr = sc_can_get_idr(frame->id, (frame->flags & CAN_FRAME_IDE), (frame->flags & CAN_FRAME_RTR));
+	sc_can_write32(config, SCCAN_TMR1_OFFSET, idr);
+
+	/* Write DLC to TMR2 */
+	sc_can_write32(config, SCCAN_TMR2_OFFSET, frame->dlc);
+
+	/* Write CAN Data Frame to TMR3/TMR4 */
+	sc_can_write32(config, SCCAN_TMR3_OFFSET, sys_be32_to_cpu(frame->data_32[0]));
+	sc_can_write32(config, SCCAN_TMR4_OFFSET, sys_be32_to_cpu(frame->data_32[1]));
+
+	/* Save call back function */
+	/* TODO:
+	 * Current SC CAN Controller transmits CAN packets in the order written
+	 * to the TX buffer from user, so call back pointer is stored on the
+	 * premise of FIFO.
+	 * However we plan to improve the sending algorithm on SC CAN controller.
+	 */
+	data->tx_cb[data->tx_head] = callback;
+	data->tx_cb_arg[data->tx_head] = user_data;
+	data->tx_head++;
+	if (data->tx_head == SCCAN_TX_FIFO_DEPTH) {
+		data->tx_head  = 0;
+	}
+
+	k_mutex_unlock(&data->inst_mutex);
+
+	return 0;
+}
+
+/*
+ * SC CAN controller has four "Acceptance filter" feature, so this driver use
+ * it as the "RX filter".
+ * And, due to the specifications of the SC CAN controller, Acceptance Filter
+ * can only be set when disable CAN.
+ * However, Zephyr specification don't defines the timing for adding/removing
+ * RX filter. In fact, CAN sample application (samples/drivers/can/counter) is
+ * called can_add_rx_filter_msgq() after can_start().
+ * Therefore, in this driver, temporarily disabled CAN and set
+ * "Acceptance filer" to SC CAN controller.
+ */
+static int sc_can_add_rx_filter(const struct device *dev, can_rx_callback_t cb,
+				  void *cb_arg, const struct can_filter *filter)
+{
+	const struct sc_can_cfg *config = dev->config;
+	struct sc_can_data *data = dev->data;
+	bool do_disable = false;
+	int ret = 0;
+	int filter_id;
+
+	LOG_DBG("Setting filter ID: 0x%x, mask: 0x%x", filter->id, filter->mask);
+
+	k_mutex_lock(&data->inst_mutex, K_FOREVER);
+
+	if (sc_can_is_enabled(config)) {
+		/* Disable CAN with retry if enabled */
+		ret = sc_can_disable(config);
+		if (ret != 0) {
+			goto unlock;
+		}
+		do_disable = true;
```

Comment:
`do_disable` sounds like we will disable it.  But the flag is used to mean "we've disabled".  So the flag should be named `disabled`, as you commented when you re-enable it.

```c
if (disabled) {
    sc_can_enable(...);
}
```

Or, `need_enable` or `reenable` is also a good candidate. so that 

```c
if (need_enable)  {
    sc_can_enable(...);
}
```

Reply:
Thanks. as you said, this flag’s name is not correctly.
Will change to `need_enable`.

Reply:
you still have it in `sc_can_remove_rx_filter()`.